File: skills/install-slop-guard/assets/python/rules/smells/no_refused_bequest.py
"""Reject subclasses that inherit a method only to refuse it."""

from __future__ import annotations

import ast

from ...rule import FunctionNode, Rule

ABSTRACT_BASES = frozenset({"ABC", "ABCMeta", "Protocol"})


class NoRefusedBequest(Rule):
    name = "no-refused-bequest"
    group = "smells"
    reference = "https://refactoring.guru/smells/refused-bequest"
    description = "Disallow overriding an inherited method with `raise NotImplementedError`."

    def visit_ClassDef(self, node: ast.ClassDef) -> None:
        if not node.bases or self._is_abstract(node):
            self.generic_visit(node)
            return
        for statement in node.body:
            if isinstance(statement, (ast.FunctionDef, ast.AsyncFunctionDef)) and self._refuses(statement):
                base = self.unparse(node.bases[0])
                self.report(
                    statement,
                    f"`{node.name}` inherits `{statement.name}` from `{base}` and then refuses "
                    f"it, so it is not the thing its base class promises. Take what you need by "
                    f"holding a `{base}` instead of inheriting from it.",
                )
        self.generic_visit(node)

    def _is_abstract(self, node: ast.ClassDef) -> bool:
        if any(self.unparse(base).split(".")[-1] in ABSTRACT_BASES for base in node.bases):
            return True
        return any(
            self.unparse(keyword.value).split(".")[-1] in ABSTRACT_BASES
            for keyword in node.keywords
            if keyword.arg == "metaclass"
        )

    def _refuses(self, node: FunctionNode) -> bool:
        if any("abstract" in self.unparse(decorator) for decorator in node.decorator_list):
            return False
        body = [
            statement
            for statement in node.body
            if not (
                isinstance(statement, ast.Expr)
                and isinstance(statement.value, ast.Constant)
                and isinstance(statement.value.value, str)
            )
        ]
        if len(body) != 1 or not isinstance(body[0], ast.Raise):
            return False
        raised = body[0].exc
        name = self.unparse(raised) if raised is not None else ""
        return name.split("(")[0].split(".")[-1] == "NotImplementedError"

File: skills/install-slop-guard/assets/typescript/smells/rules/no-refused-bequest.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

function refusesInheritance(node: ESTree.MethodDefinition): boolean {
	const body = node.value.body;
	if (body === null || body === undefined || body.type !== "BlockStatement") return false;
	if (body.body.length !== 1) return false;
	const only = body.body[0];
	return only?.type === "ThrowStatement";
}

/** Ban subclasses that inherit a method only to refuse it. */
export const noRefusedBequestRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow overriding an inherited method with a bare `throw`; a subclass that refuses its base class is not that base class.",
		},
		messages: {
			refusedBequest:
				"`{{method}}` is inherited from `{{base}}` and then refused, so this class is not the thing its base class promises. Take what you need by holding a `{{base}}` instead of extending it.",
		},
	},
	createOnce(context) {
		return {
			ClassBody(node) {
				const parent = node.parent;
				if (parent?.type !== "ClassDeclaration" && parent?.type !== "ClassExpression") return;
				const superClass = parent.superClass;
				if (superClass === null || superClass === undefined) return;
				if (parent.abstract === true) return;
				for (const member of node.body) {
					if (member.type !== "MethodDefinition" || member.kind === "constructor") continue;
					if (!refusesInheritance(member)) continue;
					context.report({
						node: member,
						messageId: "refusedBequest",
						data: {
							method:
								member.key.type === "Identifier" ? member.key.name : context.sourceCode.getText(member.key),
							base: context.sourceCode.getText(superClass),
						},
					});
				}
			},
		};
	},
});
